File: windows/src/HoverPocket.Shell/Capabilities/CapabilityBrokerModels.cs
using System.Globalization;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;

namespace HoverPocket.Shell.Capabilities;

internal enum CapabilityOrigin
{
    NativeUi,
    Voice,
    Text,
    PocketSurface,
    Mcp,
    Connector
}

internal static class CapabilityOriginExtensions
{
    public static string WireValue(this CapabilityOrigin origin) => origin switch
    {
        CapabilityOrigin.NativeUi => "native_ui",
        CapabilityOrigin.Voice => "voice",
        CapabilityOrigin.Text => "text",
        CapabilityOrigin.PocketSurface => "pocket_surface",
        CapabilityOrigin.Mcp => "mcp",
        CapabilityOrigin.Connector => "connector",
        _ => throw new ArgumentOutOfRangeException(nameof(origin))
    };
}

internal sealed record CapabilityPrincipal(
    string UserId,
    string? PocketAppId = null,
    string? AgentSessionId = null);

internal sealed record CapabilityAppContext(
    string Id,
    string Version,
    string ManifestDigest);

internal enum CapabilityEffect
{
    Pure,
    PrivateRead,
    ReversibleLocalWrite,
    ExternalWrite,
    DestructiveSensitive,
    NativeAuthority
}

internal static class CapabilityEffectExtensions
{
    public static bool IsWrite(this CapabilityEffect effect) => effect is not (CapabilityEffect.Pure or CapabilityEffect.PrivateRead);

    public static string WireValue(this CapabilityEffect effect) => effect switch
    {
        CapabilityEffect.Pure => "pure",
        CapabilityEffect.PrivateRead => "private_read",
        CapabilityEffect.ReversibleLocalWrite => "reversible_local_write",
        CapabilityEffect.ExternalWrite => "external_write",
        CapabilityEffect.DestructiveSensitive => "destructive_sensitive",
        CapabilityEffect.NativeAuthority => "native_authority",
        _ => throw new ArgumentOutOfRangeException(nameof(effect))
    };
}

internal enum CapabilityApprovalPolicy
{
    None,
    PermissionGrant,
    BrokerPolicy,
    PerCall,
    StrongPerCall,
    RuntimeProhibited
}

internal static class CapabilityApprovalPolicyExtensions
{
    public static bool RequiresExecutionApproval(this CapabilityApprovalPolicy policy) =>
        policy is CapabilityApprovalPolicy.BrokerPolicy or CapabilityApprovalPolicy.PerCall or CapabilityApprovalPolicy.StrongPerCall;

    public static string WireValue(this CapabilityApprovalPolicy policy) => policy switch
    {
        CapabilityApprovalPolicy.None => "none",
        CapabilityApprovalPolicy.PermissionGrant => "permission_grant",
        CapabilityApprovalPolicy.BrokerPolicy => "broker_policy",
        CapabilityApprovalPolicy.PerCall => "per_call",
        CapabilityApprovalPolicy.StrongPerCall => "strong_per_call",
        CapabilityApprovalPolicy.RuntimeProhibited => "runtime_prohibited",
        _ => throw new ArgumentOutOfRangeException(nameof(policy))
    };
}

internal enum CapabilityIdempotencyPolicy
{
    NotApplicable,
    Optional,
    Required
}

internal enum CapabilityReadbackStrategy
{
    None,
    EntityGetById,
    CapabilityQuery,
    SameStoreSnapshot,
    OsState,
    ContentDigest
}

internal static class CapabilityReadbackStrategyExtensions
{
    public static string WireValue(this CapabilityReadbackStrategy strategy) => strategy switch
    {
        CapabilityReadbackStrategy.None => "none",
        CapabilityReadbackStrategy.EntityGetById => "entity_get_by_id",
        CapabilityReadbackStrategy.CapabilityQuery => "capability_query",
        CapabilityReadbackStrategy.SameStoreSnapshot => "same_store_snapshot",
        CapabilityReadbackStrategy.OsState => "os_state",
        CapabilityReadbackStrategy.ContentDigest => "content_digest",
        _ => throw new ArgumentOutOfRangeException(nameof(strategy))
    };
}

internal sealed record CapabilityLimits(
    int TimeoutMilliseconds,
    int MaximumPayloadBytes,
    int MaximumCallsPerMinute);

internal sealed record CapabilityReadbackPolicy(
    CapabilityReadbackStrategy Strategy,
    PocketCapabilityKey? Query,
    IReadOnlyList<string> MatchFields);

internal sealed record CapabilityPlanStep(
    string Id,
    PocketCapabilityKey Capability,
    JsonElement Arguments,
    string IdempotencyKey,
    IReadOnlyList<string> Dependencies);

internal sealed record CapabilityExecutionPlan(
    string Id,
    DateTimeOffset CreatedAt,
    CapabilityOrigin Origin,
    CapabilityPrincipal Principal,
    CapabilityAppContext? AppContext,
    IReadOnlyList<CapabilityPlanStep> Steps,
    IReadOnlySet<string> RequiredPermissions);

internal sealed record CapabilityPermissionSet(
    CapabilityPrincipal Principal,
    IReadOnlySet<string> Permissions)
{
    public bool Contains(IReadOnlySet<string> required) => required.All(Permissions.Contains);
}

internal sealed record CapabilityApprovalEffect(
    string StepId,
    PocketCapabilityKey Capability,
    CapabilityEffect Effect,
    string ArgumentDigest,
    string SummaryKey,
    bool RollbackAvailable);

internal sealed record CapabilityApprovalRequest(
    string Id,
    string PlanId,
    string PlanDigest,
    CapabilityPrincipal Principal,
    CapabilityAppContext? AppContext,
    DateTimeOffset CreatedAt,
    DateTimeOffset ExpiresAt,
    string Nonce,
    IReadOnlyList<CapabilityApprovalEffect> Effects,
    IReadOnlySet<string> RequiredPermissions);

internal readonly record struct CapabilityApprovalGrant(string Token);

internal sealed record CapabilityBrokerPreparation(
    string PlanDigest,
    CapabilityApprovalRequest? ApprovalRequest);

internal enum CapabilityReceiptStatus
{
    Succeeded,
    Rejected,
    Failed,
    Partial,
    Unknown
}

internal static class CapabilityReceiptStatusExtensions
{
    public static string WireValue(this CapabilityReceiptStatus status) => status.ToString().ToLowerInvariant();
}

internal enum CapabilityReadbackStatus
{
    Verified,
    Mismatch,
    Unavailable
}

internal static class CapabilityReadbackStatusExtensions
{
    public static string WireValue(this CapabilityReadbackStatus status) => status.ToString().ToLowerInvariant();
}

internal sealed record CapabilityReadbackReceipt(
    CapabilityReadbackStatus Status,
    CapabilityReadbackStrategy Strategy,
    DateTimeOffset? ObservedAt,
    JsonElement? Observed,
    string? EvidenceDigest);

internal sealed record CapabilitySafeError(
    string Code,
    bool Retryable,
    string MessageKey);

internal sealed record CapabilityReceipt(
    string InvocationId,
    string PlanId,
    string PlanDigest,
    PocketCapabilityKey Capability,
    CapabilityReceiptStatus Status,
    JsonElement? Output,
    CapabilityReadbackReceipt Readback,
    bool RollbackAvailable,
    string? RollbackStatus,
    string AuditEntryId,
    CapabilitySafeError? SafeError,
    DateTimeOffset CompletedAt,
    bool Replayed)
{
    public CapabilityReceipt ReplayCopy() => this with { Replayed = true };
}

internal sealed record CapabilityWorkflowReceipt(
    string PlanId,
    string PlanDigest,
    CapabilityReceiptStatus Status,
    IReadOnlyList<CapabilityReceipt> Steps,
    DateTimeOffset CompletedAt,
    bool Replayed)
{
    public CapabilityWorkflowReceipt ReplayCopy() => this with
    {
        Steps = Steps.Select(receipt => receipt.ReplayCopy()).ToArray(),
        Replayed = true
    };
}

internal sealed class CapabilityBrokerException(string code, string field) : Exception($"{code}: {field}")
{
    public string Code { get; } = code;
    public string Field { get; } = field;
}

internal static class CapabilityCanonicalJson
{
    public static string ArgumentsDigest(JsonElement arguments) => Digest(CanonicalBytes(arguments));

    public static string PlanDigest(CapabilityExecutionPlan plan)
    {
        using var stream = new MemoryStream();
        using (var writer = new Utf8JsonWriter(stream))
        {
            writer.WriteStartObject();
            if (plan.AppContext is not null)
            {
                writer.WritePropertyName("appContext");
                WriteAppContext(writer, plan.AppContext);
            }
            writer.WriteString("createdAt", Date(plan.CreatedAt));
            writer.WriteString("origin", plan.Origin.WireValue());
            writer.WriteString("planId", plan.Id);
            writer.WriteNumber("planVersion", 1);
            writer.WritePropertyName("principal");
            WritePrincipal(writer, plan.Principal);
            writer.WritePropertyName("requiredPermissions");
            writer.WriteStartArray();
            foreach (var permission in plan.RequiredPermissions.Order(StringComparer.Ordinal))
            {
                writer.WriteStringValue(permission);
            }
            writer.WriteEndArray();
            writer.WritePropertyName("steps");
            writer.WriteStartArray();
            foreach (var step in plan.Steps)
            {
                writer.WriteStartObject();
                writer.WritePropertyName("arguments");
                WriteElement(writer, step.Arguments);
                writer.WriteString("capabilityId", step.Capability.Id);
                writer.WriteNumber("capabilityVersion", step.Capability.Version);
                writer.WritePropertyName("dependsOn");
                writer.WriteStartArray();
                foreach (var dependency in step.Dependencies.Order(StringComparer.Ordinal))
                {
                    writer.WriteStringValue(dependency);
                }
                writer.WriteEndArray();
                writer.WriteString("idempotencyKey", step.IdempotencyKey);
                writer.WriteString("stepId", step.Id);
                writer.WriteEndObject();
            }
            writer.WriteEndArray();
            writer.WriteEndObject();
        }
        return Digest(stream.ToArray());
    }

    public static byte[] CanonicalBytes(JsonElement element)
    {
        using var stream = new MemoryStream();
        using (var writer = new Utf8JsonWriter(stream))
        {
            WriteElement(writer, element);
        }
        return stream.ToArray();
    }

    public static string Date(DateTimeOffset date) =>
        date.ToUniversalTime().ToString("yyyy-MM-dd'T'HH:mm:ss.fff'Z'", CultureInfo.InvariantCulture);

    private static string Digest(byte[] bytes) =>
        "sha256:" + Convert.ToHexString(SHA256.HashData(bytes)).ToLowerInvariant();

    private static void WritePrincipal(Utf8JsonWriter writer, CapabilityPrincipal principal)
    {
        writer.WriteStartObject();
        if (principal.AgentSessionId is not null)
        {
            writer.WriteString("agentSessionId", principal.AgentSessionId);
        }
        if (principal.PocketAppId is not null)
        {
            writer.WriteString("pocketAppId", principal.PocketAppId);
        }
        writer.WriteString("userId", principal.UserId);
        writer.WriteEndObject();
    }

    private static void WriteAppContext(Utf8JsonWriter writer, CapabilityAppContext app)
    {
        writer.WriteStartObject();
        writer.WriteString("id", app.Id);
        writer.WriteString("manifestDigest", app.ManifestDigest);
        writer.WriteString("version", app.Version);
        writer.WriteEndObject();
    }

    public static void WriteElement(Utf8JsonWriter writer, JsonElement element)
    {
        switch (element.ValueKind)
        {
            case JsonValueKind.Object:
                writer.WriteStartObject();
                foreach (var property in element.EnumerateObject().OrderBy(property => property.Name, StringComparer.Ordinal))
                {
                    writer.WritePropertyName(property.Name);
                    WriteElement(writer, property.Value);
                }
                writer.WriteEndObject();
                break;
            case JsonValueKind.Array:
                writer.WriteStartArray();
                foreach (var item in element.EnumerateArray())
                {
                    WriteElement(writer, item);
                }
                writer.WriteEndArray();
                break;
            case JsonValueKind.String:
                writer.WriteStringValue(element.GetString());
                break;
            case JsonValueKind.Number:
                element.WriteTo(writer);
                break;
            case JsonValueKind.True:
            case JsonValueKind.False:
                writer.WriteBooleanValue(element.GetBoolean());
                break;
            case JsonValueKind.Null:
                writer.WriteNullValue();
                break;
            default:
                throw new CapabilityBrokerException("CAPABILITY_ARGUMENT_INVALID", "json");
        }
    }
}

File: Sources/HoverPocket/Capabilities/CapabilityBrokerModels.swift
import CryptoKit
import Foundation

extension CapabilityValue: Codable {
    init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        if container.decodeNil() {
            self = .null
        } else if let value = try? container.decode(Bool.self) {
            self = .bool(value)
        } else if let value = try? container.decode(Int.self) {
            self = .integer(value)
        } else if let value = try? container.decode(Double.self) {
            guard value.isFinite else {
                throw DecodingError.dataCorruptedError(in: container, debugDescription: "Non-finite number")
            }
            self = .number(value)
        } else if let value = try? container.decode(String.self) {
            self = .string(value)
        } else if let value = try? container.decode([CapabilityValue].self) {
            self = .array(value)
        } else {
            self = .object(try container.decode([String: CapabilityValue].self))
        }
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        switch self {
        case .null:
            try container.encodeNil()
        case .bool(let value):
            try container.encode(value)
        case .integer(let value):
            try container.encode(value)
        case .number(let value):
            try container.encode(value)
        case .string(let value):
            try container.encode(value)
        case .array(let value):
            try container.encode(value)
        case .object(let value):
            try container.encode(value)
        }
    }
}

enum CapabilityOrigin: String, Codable, CaseIterable, Sendable {
    case nativeUI = "native_ui"
    case voice
    case text
    case pocketSurface = "pocket_surface"
    case mcp
    case connector
}

struct CapabilityPrincipal: Codable, Equatable, Sendable {
    let userID: String
    let pocketAppID: String?
    let agentSessionID: String?

    init(userID: String, pocketAppID: String? = nil, agentSessionID: String? = nil) {
        self.userID = userID
        self.pocketAppID = pocketAppID
        self.agentSessionID = agentSessionID
    }
}

struct CapabilityAppContext: Codable, Equatable, Sendable {
    let id: String
    let version: String
    let manifestDigest: String
}

enum CapabilityEffect: String, Codable, Sendable {
    case pure
    case privateRead = "private_read"
    case reversibleLocalWrite = "reversible_local_write"
    case externalWrite = "external_write"
    case destructiveSensitive = "destructive_sensitive"
    case nativeAuthority = "native_authority"

    var isWrite: Bool {
        switch self {
        case .pure, .privateRead:
            false
        case .reversibleLocalWrite, .externalWrite, .destructiveSensitive, .nativeAuthority:
            true
        }
    }
}

enum CapabilityApprovalPolicy: String, Codable, Sendable {
    case none
    case permissionGrant = "permission_grant"
    case brokerPolicy = "broker_policy"
    case perCall = "per_call"
    case strongPerCall = "strong_per_call"
    case runtimeProhibited = "runtime_prohibited"

    var requiresExecutionApproval: Bool {
        switch self {
        case .brokerPolicy, .perCall, .strongPerCall:
            true
        case .none, .permissionGrant, .runtimeProhibited:
            false
        }
    }
}

enum CapabilityIdempotencyPolicy: String, Codable, Sendable {
    case notApplicable = "not_applicable"
    case optional
    case required
}

enum CapabilityReadbackStrategy: String, Codable, Sendable {
    case none
    case entityGetByID = "entity_get_by_id"
    case capabilityQuery = "capability_query"
    case sameStoreSnapshot = "same_store_snapshot"
    case osState = "os_state"
    case contentDigest = "content_digest"
}

struct CapabilityLimits: Codable, Equatable, Sendable {
    let timeoutMilliseconds: Int
    let maximumPayloadBytes: Int
    let maximumCallsPerMinute: Int
}

struct CapabilityReadbackPolicy: Codable, Equatable, Sendable {
    let strategy: CapabilityReadbackStrategy
    let query: PocketCapabilityKey?
    let matchFields: [String]
}

struct CapabilityPlanStep: Codable, Equatable, Sendable {
    let id: String
    let capability: PocketCapabilityKey
    let arguments: CapabilityObject
    let idempotencyKey: String
    let dependencies: [String]
}

struct CapabilityExecutionPlan: Codable, Equatable, Sendable {
    let id: String
    let createdAt: Date
    let origin: CapabilityOrigin
    let principal: CapabilityPrincipal
    let appContext: CapabilityAppContext?
    let steps: [CapabilityPlanStep]
    let requiredPermissions: Set<String>
}

struct CapabilityPermissionSet: Equatable, Sendable {
    let principal: CapabilityPrincipal
    let permissions: Set<String>

    func contains(_ required: Set<String>) -> Bool {
        required.isSubset(of: permissions)
    }
}

struct CapabilityApprovalEffect: Codable, Equatable, Sendable {
    let stepID: String
    let capability: PocketCapabilityKey
    let effect: CapabilityEffect
    let argumentDigest: String
    let summaryKey: String
    let rollbackAvailable: Bool
}

struct CapabilityApprovalRequest: Codable, Equatable, Sendable {
    let id: String
    let planID: String
    let planDigest: String
    let principal: CapabilityPrincipal
    let appContext: CapabilityAppContext?
    let createdAt: Date
    let expiresAt: Date
    let nonce: String
    let effects: [CapabilityApprovalEffect]
    let requiredPermissions: Set<String>
}

struct CapabilityBrokerPreparation: Equatable, Sendable {
    let planDigest: String
    let approvalRequest: CapabilityApprovalRequest?
    let approvalPresentations: [CapabilityApprovalPresentation]
}

struct CapabilityApprovalGrant: Equatable, Sendable {
    let token: String
}

enum CapabilityReceiptStatus: String, Codable, Sendable {
    case succeeded
    case rejected
    case failed
    case partial
    case unknown
}

enum CapabilityReadbackStatus: String, Codable, Sendable {
    case verified
    case mismatch
    case unavailable
}

struct CapabilityReadbackReceipt: Codable, Equatable, Sendable {
    let status: CapabilityReadbackStatus
    let strategy: CapabilityReadbackStrategy
    let observedAt: Date?
    let observed: CapabilityObject?
    let evidenceDigest: String?
}

struct CapabilitySafeError: Codable, Equatable, Sendable {
    let code: String
    let retryable: Bool
    let messageKey: String
}

struct CapabilityReceipt: Codable, Equatable, Sendable {
    let invocationID: String
    let planID: String
    let planDigest: String
    let capability: PocketCapabilityKey
    let status: CapabilityReceiptStatus
    let output: CapabilityObject?
    let readback: CapabilityReadbackReceipt
    let rollbackAvailable: Bool
    let rollbackStatus: String?
    let auditEntryID: String
    let safeError: CapabilitySafeError?
    let completedAt: Date
    let replayed: Bool

    func replayCopy() -> CapabilityReceipt {
        CapabilityReceipt(
            invocationID: invocationID,
            planID: planID,
            planDigest: planDigest,
            capability: capability,
            status: status,
            output: output,
            readback: readback,
            rollbackAvailable: rollbackAvailable,
            rollbackStatus: rollbackStatus,
            auditEntryID: auditEntryID,
            safeError: safeError,
            completedAt: completedAt,
            replayed: true
        )
    }

    func durableCopy() -> CapabilityReceipt {
        CapabilityReceipt(
            invocationID: invocationID,
            planID: planID,
            planDigest: planDigest,
            capability: capability,
            status: status,
            output: nil,
            readback: CapabilityReadbackReceipt(
                status: readback.status,
                strategy: readback.strategy,
                observedAt: readback.observedAt,
                observed: nil,
                evidenceDigest: readback.evidenceDigest
            ),
            rollbackAvailable: rollbackAvailable,
            rollbackStatus: rollbackStatus,
            auditEntryID: auditEntryID,
            safeError: safeError,
            completedAt: completedAt,
            replayed: replayed
        )
    }

    func withRollbackStatus(_ status: String) -> CapabilityReceipt {
        CapabilityReceipt(
            invocationID: invocationID,
            planID: planID,
            planDigest: planDigest,
            capability: capability,
            status: self.status,
            output: output,
            readback: readback,
            rollbackAvailable: rollbackAvailable,
            rollbackStatus: status,
            auditEntryID: auditEntryID,
            safeError: safeError,
            completedAt: completedAt,
            replayed: replayed
        )
    }
}

struct CapabilityWorkflowReceipt: Codable, Equatable, Sendable {
    let planID: String
    let planDigest: String
    let status: CapabilityReceiptStatus
    let steps: [CapabilityReceipt]
    let completedAt: Date
    let replayed: Bool

    func replayCopy() -> CapabilityWorkflowReceipt {
        CapabilityWorkflowReceipt(
            planID: planID,
            planDigest: planDigest,
            status: status,
            steps: steps.map { $0.replayCopy() },
            completedAt: completedAt,
            replayed: true
        )
    }

    func durableCopy() -> CapabilityWorkflowReceipt {
        CapabilityWorkflowReceipt(
            planID: planID,
            planDigest: planDigest,
            status: status,
            steps: steps.map { $0.durableCopy() },
            completedAt: completedAt,
            replayed: replayed
        )
    }
}

enum CapabilityBrokerError: Error, Equatable, Sendable {
    case invalidPlan(String)
    case unknownCapability(PocketCapabilityKey)
    case removedCapability(PocketCapabilityKey, PocketCapabilityKey?)
    case unavailable(PocketCapabilityKey)
    case runtimeProhibited(PocketCapabilityKey)
    case invalidArguments(PocketCapabilityKey, String)
    case permissionDenied(String)
    case approvalRequired
    case approvalRejected
    case approvalExpired
    case approvalInvalid
    case approvalReplayed
    case idempotencyConflict(String)
    case executionUnknown(String)
    case rateLimited(PocketCapabilityKey)
    case timedOut(PocketCapabilityKey)
    case ledgerUnavailable
}

enum CapabilityCanonicalJSON {
    static func digest(_ value: CapabilityValue) throws -> String {
        let data = try data(value)
        return "sha256:" + SHA256.hash(data: data).map { String(format: "%02x", $0) }.joined()
    }

    static func digest(_ object: CapabilityObject) throws -> String {
        try digest(.object(object))
    }

    static func data(_ value: CapabilityValue) throws -> Data {
        try JSONSerialization.data(withJSONObject: foundation(value), options: [.sortedKeys, .withoutEscapingSlashes])
    }

    static func planDigest(_ plan: CapabilityExecutionPlan) throws -> String {
        let steps = plan.steps.map { step in
            CapabilityValue.object([
                "arguments": .object(step.arguments),
                "capabilityId": .string(step.capability.id),
                "capabilityVersion": .integer(step.capability.version),
                "dependsOn": .array(step.dependencies.sorted().map(CapabilityValue.string)),
                "idempotencyKey": .string(step.idempotencyKey),
                "stepId": .string(step.id)
            ])
        }
        var principal: CapabilityObject = ["userId": .string(plan.principal.userID)]
        if let value = plan.principal.pocketAppID { principal["pocketAppId"] = .string(value) }
        if let value = plan.principal.agentSessionID { principal["agentSessionId"] = .string(value) }
        var root: CapabilityObject = [
            "createdAt": .string(CapabilityDateCodec.string(from: plan.createdAt)),
            "origin": .string(plan.origin.rawValue),
            "planId": .string(plan.id),
            "planVersion": .integer(1),
            "principal": .object(principal),
            "requiredPermissions": .array(plan.requiredPermissions.sorted().map(CapabilityValue.string)),
            "steps": .array(steps)
        ]
        if let app = plan.appContext {
            root["appContext"] = .object([
                "id": .string(app.id),
                "manifestDigest": .string(app.manifestDigest),
                "version": .string(app.version)
            ])
        }
        return try digest(root)
    }

    private static func foundation(_ value: CapabilityValue) -> Any {
        switch value {
        case .null:
            NSNull()
        case .bool(let value):
            value
        case .integer(let value):
            value
        case .number(let value):
            value
        case .string(let value):
            value
        case .array(let values):
            values.map(foundation)
        case .object(let object):
            object.mapValues(foundation)
        }
    }
}

extension PocketCapabilityKey: Codable {
    private enum CodingKeys: String, CodingKey {
        case id
        case version
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        self.init(
            id: try container.decode(String.self, forKey: .id),
            version: try container.decode(Int.self, forKey: .version)
        )
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(id, forKey: .id)
        try container.encode(version, forKey: .version)
    }
}
